src/asmatrix.c

```diff
@@ -0,0 +1,257 @@
+#include "data.table.h"
+
+/* To handle different matrix types (e.g. numeric, character, etc)
+ * there is a single dispatch function asmatrix at the bottom of the 
+ * file that detects the appropriate R atomic type then calls the
+ * corresponding asmatrix_<type> function.
+ */
+
+SEXP asmatrix_logical(SEXP dt, R_xlen_t matlen, R_xlen_t n, R_xlen_t rncolnum) {
+  // Create output matrix, allocate memory, and create a pointer to it.
+  SEXP mat = PROTECT(allocVector(LGLSXP, matlen));
+  int *pmat; 
+  pmat = LOGICAL(mat);
+  
+  /* Create an array of pointers for the individual columns in DT 
+   * so that we avoid creating these and calling the helper functions
+   * in the parallel OMP threads
+   */
+  R_xlen_t dtncol = xlength(dt);
```

Comment:
Is it impossible to have more than 2^31 - 1 columns in a data.table?

Reply:
I doubt, even if it is, then I don't think we should really care about support of such.

Reply:
i don't see support of this in any similar tools... Vowpal Wabbit is able to support it technically but I don't know of any use case

https://github.com/VowpalWabbit/vowpal_wabbit/wiki

> ...it can be effectively applied on learning problems with a sparse terafeature (i.e. 10^12 sparse features)

emphasis on sparse, not supported by `data.table`, not sure we have any plan to

Reply:
Actually it looks like we should change both dimensions to R_len_t. 

From the R internals page:

> Matrices are stored as vectors and so were also limited to 2^31-1 elements. Now longer vectors are allowed on 64-bit platforms, matrices with more elements are supported *provided that each of the dimensions is no more than 2^31-1*. However, not all applications can be supported.

Reply:
There is no concept of sparse columns because they are not aligned in memory. Each column is another pointer to new mem location. Sparse columns make sense for matrix only. I think the best to limit each dimension to what R internals suggests.

Reply:
I've updated the C code to use R_len_t where possible, and check for dimensions exceeding that in the dispatch function. I've had to keep the variable storing the number of rows (`n`) as type `R_xlen_t` as its used to increment the vector index counter (`vecIdx`) which must be type `R_xlen_t` to accomodate matrices whose total length is > 2^31 - 1.